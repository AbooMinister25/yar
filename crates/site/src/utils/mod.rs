use std::path::{Path, PathBuf};

use color_eyre::{Result, eyre::ContextCompat};
use url::Url;

pub mod fs;

/// Build permalink for a site item.
pub fn build_permalink<P: AsRef<Path>, T: AsRef<Path>>(
    path: P,
    out_dir: T,
    url: &Url,
) -> Result<Url> {
    let out = out_dir
        .as_ref()
        .file_name()
        .context("Output directory shouldn't terminate in ..")?;

    let mut components = path
        .as_ref()
        .components()
        .skip_while(|c| c.as_os_str() != out);
    components.next();
    let mut url_ending = components.collect::<PathBuf>();
    if url_ending.ends_with("index.html") {
        url_ending = url_ending
            .parent()
            .context("path doesn't have parent?")?
            .to_path_buf();
    }
    let permalink = url.join(
        url_ending
            .to_str()
            .context("Path should be valid unicode.")?,
    )?;

    Ok(permalink)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_build_permalink() -> Result<()> {
        let path = Path::new("site/index.html");
        let out_dir = Path::new("site/");
        let url = Url::parse("https://example.com/")?;
        insta::assert_yaml_snapshot!(build_permalink(path, out_dir, &url)?);

        let path = Path::new("site/posts/hello-world/index.html");
        let out_dir = Path::new("site/");
        let url = Url::parse("https://example.com/")?;
        insta::assert_yaml_snapshot!(build_permalink(path, out_dir, &url)?);

        let path = Path::new("site/assets/style.css");
        let out_dir = Path::new("site/");
        let url = Url::parse("https://example.com/")?;
        insta::assert_yaml_snapshot!(build_permalink(path, out_dir, &url)?);

        Ok(())
    }
}
