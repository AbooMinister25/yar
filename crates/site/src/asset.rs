use std::{
    ffi::OsStr,
    fs,
    path::{Path, PathBuf},
};

use color_eyre::{Result, eyre::ContextCompat};
use url::Url;

use crate::utils::{build_permalink, fs::ensure_directory};

/// Represents a resource that is passed through an asset pipeline.
/// This can include things like images, stylesheets, and javascript.
#[derive(Debug)]
pub struct Asset {
    pub path: PathBuf,
    pub source_hash: String,
    pub out_path: PathBuf,
    pub permalink: Url,
    pub content: String,
}

impl Asset {
    pub fn new<P: AsRef<Path>, T: AsRef<Path>, Z: AsRef<Path>>(
        path: P,
        source_hash: String,
        out_dir: T,
        root: Z,
        url: &Url,
    ) -> Result<Self> {
        let out_path = out_path(&path, &out_dir, root);
        let (content, out_path) = process_asset(&path, out_path)?;
        let permalink = build_permalink(&out_path, out_dir, url)?;

        Ok(Self {
            path: path.as_ref().to_owned(),
            source_hash,
            out_path,
            permalink,
            content,
        })
    }

    pub fn render(&self) -> Result<()> {
        ensure_directory(
            self.out_path
                .parent()
                .context("Path should have a parent")?,
        )?;
        fs::write(&self.out_path, &self.content)?;
        Ok(())
    }
}

fn process_asset<P: AsRef<Path>, T: AsRef<Path>>(path: P, out_dir: T) -> Result<(String, PathBuf)> {
    let mut op = out_dir.as_ref().to_owned();
    let options = grass::Options::default().style(grass::OutputStyle::Compressed);

    Ok((
        match path.as_ref().extension().and_then(OsStr::to_str) {
            Some("scss") => {
                op.set_extension("css");
                grass::from_path(path, &options)?
            }
            Some(ext) => {
                op.set_extension(ext);
                fs::read_to_string(path)?
            }
            None => fs::read_to_string(path)?,
        },
        op,
    ))
}

fn out_path<P: AsRef<Path>, T: AsRef<Path>, Z: AsRef<Path>>(
    path: P,
    out_dir: T,
    root: Z,
) -> PathBuf {
    let out_dir = out_dir.as_ref();
    let mut components = path
        .as_ref()
        .components()
        .filter(|c| !c.as_os_str().to_str().is_some_and(|s| s.starts_with("_")));

    if root.as_ref() != Path::new(".") {
        components.next();
    }

    out_dir.components().chain(components).collect::<PathBuf>()
}
