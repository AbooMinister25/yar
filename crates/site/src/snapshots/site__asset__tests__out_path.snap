---
source: crates/site/src/asset.rs
expression: path
---
public/assets/style.scss
